### DISCUSSION TITLE
Multiple layers of variadic route path components

### DISCUSSION THREAD

**Discussion by edgarogh** [tokio-rs/axum#3008] [Q&A] [answered]
### Summary

Hello, I'd like to define a router that supports a variadic number of path components up to a specific delimiter (`/-/` like GitLab does), then within that, I'd like to nest a second router that may also have sub-routes with variadic path components.

This looks exceptionally hard to do in axum since path components are not consumed by parent routers. So even with a custom serde implementation of a sequence visitor for the path extractor, that stops when it encounters the `/-/`, the inner router would still parse the path from the very beginning. I'm on the verge of re-implementing the entire router myself because it's really to cumbersome to work with...

Is there any way to work around this?

### axum version

0.7.7

**Comment by mladedav** [comment, marked as answer]
Do you mean [this feature](https://github.com/ibraheemdev/matchit/issues/39)? Until matchit supports it, axum cannot use variadic path components that are not at the end of the route.

I'm honestly not sure what you meant by components not being consumed by parent routers. Could you please provide an example of what you tried, what you expected and what happened?

**Reply by edgarogh** [reply]
I didn't attempt anything since the docs specifically mention that what I wanted to do isn't possible.

Basically, by "consume", I meant that `Path<T: Deserialize>` extractors could have been implemented in such a way that unnamed components (variadic components, and components extracted using tuples) are "consumed" in such a way that `Path` extractor that run after would only be able to access components from the point where the last extractor left off (unless the field has an explicit name).

By combining this with a custom `serde::Deserialize` implementation, it would've then been possible for a middleware to extract just the variadic part it is interested in, stopping at `/-/`, so that an inner router could run on the path that comes after the `/-/` delimiter.

I hope this makes more sense (even though this is clearly not possible due to the way things work). Thanks for the answer anyway. I think I'll have to reimplement a router myself.

**Comment by edgarogh** [comment]
So what I ended up doing is foul, but I literally have no other easy choice at the moment if I don't want to reimplement the router completely.

I have a top-level middleware (outside of the router) that tries to find the very specific vararg routes that cause me trouble, and rewrites them to constant routes that the router will handle, while simultaneously injecting the dynamic path parts as extensions.

**Reply by mladedav** [reply]
I think this is the simplest solution you can do right now. One small change you can consider is rewriting just slashes inside the variadic part to their percent-encoded variant. The router will then see that as a single segment and route it as such and you can extract that path in your handler without using custom extensions.

Hopefully we'll be able to support this better soon.

**Reply by edgarogh** [reply]
That's really smart, thanks for the tip :heart:
